src/tests/baseservices/finalization/CriticalFinalizer.cs

```diff
@@ -0,0 +1,72 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+// Tests a weak ordering among normal and critical finalizers: for objects reclaimed by garbage collection
+// at the same time, all the noncritical finalizers must be called before any of the critical finalizers.
+
+using System;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using System.Runtime.ConstrainedExecution;
+
+class Normal
+{
+    public static int Finalized;
+
+    ~Normal() => Finalized++;
+}
+
+class Critical : CriticalFinalizerObject
+{
+    public static int Finalized;
+    public static int NormalFinalizedBeforeFirstCritical;
+
+    ~Critical()
+    {
+        if (++Finalized == 1)
+            NormalFinalizedBeforeFirstCritical = Normal.Finalized;
+    }
+}
+
+static class CriticalFinalizerTest
+{
+    [MethodImpl(MethodImplOptions.NoInlining)]
+    static void AllocateObjects(int count)
+    {
+        List<object> list = new();
+
+        for (int i = 0; i < count; i++)
+        {
+            list.Add(new Normal());
```

Comment:
For completeness of the test, perhaps allocate on multiple threads?
It may be easier to guarantee the order if all objects end up in the same heap.

Reply:
This could be as simple as switching to Parallel.For